lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/EvalM.scala

```diff
@@ -0,0 +1,64 @@
+package com.wavesplatform.lang.v1.evaluator
+
+import cats.data.{EitherT, Kleisli}
+import cats.implicits._
+import cats.{Monad, StackSafeMonad}
+import com.wavesplatform.lang.v1.evaluator.ctx.EvaluationContext
+import com.wavesplatform.lang.{ExecutionError, TrampolinedExecResult}
+import monix.eval.Coeval
+
+final case class EvalM[A](inner: Kleisli[Coeval, CoevalRef[EvaluationContext], Either[ExecutionError, A]]) {
+  def ter(ctx: EvaluationContext): TrampolinedExecResult[A] = {
+    val atom = CoevalRef.of(ctx)
+
+    EitherT(inner.run(atom))
+  }
+
+  def run(ctx: EvaluationContext): Either[(EvaluationContext, ExecutionError), A] = {
+    val atom = CoevalRef.of(ctx)
+
+    val action = inner
+      .run(atom)
+
+    (for {
+      result  <- action
+      lastCtx <- atom.read
+    } yield result.left.map(err => (lastCtx, err))).value
+  }
+}
+
+object EvalM {
+
+  implicit val monadInstance: Monad[EvalM] = new StackSafeMonad[EvalM] {
+    override def pure[A](x: A): EvalM[A] =
+      EvalM(Kleisli.pure[Coeval, CoevalRef[EvaluationContext], Either[ExecutionError, A]](x.asRight[ExecutionError]))
+
+    override def flatMap[A, B](fa: EvalM[A])(f: A => EvalM[B]): EvalM[B] = {
```

Comment:
is it stack-safe?

Reply:
It's pass stack-safety check in tests (successful evaluation on long expression) but i'm not 100% sure , i will try to implement tailRecM later.